### PR TITLE
[GLUTEN-3582] Support PageIndex

### PR DESCRIPTION
**Pull request by baibaichen**
## What changes were proposed in this pull request?

1. Refer KeyCondition, using RPN(which is  reverse polish notation) to calculate expression. 
2. Once RowRanges is got, skip io by `getStream`
3. Still use Arrow Parquet Reader

(Fixes: \#3582)

## How was this patch tested?
Using Existed UT

### CODE REVIEW

**Comment by baibaichen** [comment]
dd7b6b232764642bcfb3abc29e70560

**Comment by baibaichen** [comment]
Run Gluten Clickhouse CI

**Comment by lwz9103** [comment]
Run Gluten Clickhouse CI

File: cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    RowRanges all() const { return RowRanges::createSingle(rowGroupRowCount); }
+
+    RowRanges toRowRanges(const PageIndexs & pages) const
+    {
+        if (pages.size() == 1 && pages[0] == PageIndexsBuilder::ALL_PAGES)
+            return all();
+        RowRanges row_ranges;
+        std::ranges::for_each(
+            pages,
+            [&](const size_t pageIndex) {
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h:96]: the `const` in `const size_t pageIndex` seems unnecessary

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+    const parquet::arrow::SchemaField & field, ParquetFileReaderExt * reader, const std::vector<int32_t> & row_groups)
+    : arrowField_(field.field)
+    , input_(field.column_index, reader, row_groups)
+    , record_reader_(parquet::internal::RecordReader::Make(
+          input_.descr(), ComputeLevelInfo(input_.descr()), default_arrow_pool(), arrowField_->type()->id() == ::arrow::Type::DICTIONARY))
+{
+    NextRowGroup();
+}
+
+void VectorizedColumnReader::NextRowGroup()
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:106]: The first letter of the name is not lowercase

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h [hunk 3]
```diff
@@ -42,9 +42,9 @@ class ParquetFormatFile : public FormatFile
 {
 public:
     explicit ParquetFormatFile(
-        DB::ContextPtr context_,
+        const DB::ContextPtr & context_,
         const substrait::ReadRel::LocalFiles::FileOrFiles & file_info_,
-        ReadBufferBuilderPtr read_buffer_builder_,
+        const ReadBufferBuilderPtr & read_buffer_builder_,
         bool useLocalFormat);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h:48]: use_local_format

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h [hunk 4]
```diff
@@ -18,27 +18,25 @@
...
 class FileReaderWrapper
 {
 public:
-    explicit FileReaderWrapper(FormatFilePtr file_) : file(file_) { }
+    explicit FileReaderWrapper(const FormatFilePtr & file_) : file(file_) { }
     virtual ~FileReaderWrapper() = default;
     virtual bool pull(DB::Chunk & chunk) = 0;
-    virtual void applyKeyCondition(std::shared_ptr<const DB::KeyCondition> /*key_condition*/) { }
+    virtual void
+    applyKeyCondition(const std::shared_ptr<const DB::KeyCondition> & /*key_condition*/, const std::shared_ptr<ColumnIndexFilter> &)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.h:37]: add a comment that ColumnIndexFilter is useless when local format is not used.

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSourceStep.cpp [hunk 5]
```diff
@@ -77,9 +69,7 @@ void SubstraitFileSourceStep::applyFilters(const DB::ActionDAGNodes added_filter
 {
     filter_actions_dag = DB::ActionsDAG::buildFilterActionsDAG(added_filter_nodes.nodes);
     for (const auto & processor : pipe.getProcessors())
-    {
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSourceStep.cpp:80]: remove bracket of for loop is not friendly for reading.

File: cpp-ch/local-engine/tests/CMakeLists.txt [hunk 6]
```diff
@@ -54,19 +54,27 @@ if (ENABLE_TESTS)
...
     # no-unreachable-code for GTEST_SKIP
     target_compile_options(unit_tests_local_engine PRIVATE -Wno-unreachable-code)
     target_link_libraries(unit_tests_local_engine PRIVATE gluten_clickhouse_backend_libs clickhouse_parsers loggers ch_contrib::gmock_all ch_contrib::gtest)
-    #TODO: target_link_libraries(unit_tests_local_engine PRIVATE ch_parquet)
-else()
+    target_link_libraries(unit_tests_local_engine PRIVATE ch_parquet)
+endif()
+
+if (ENABLE_BENCHMARKS)
     include_directories(benchmark_local_engine SYSTEM PUBLIC
```
Comment (taiyang-li) [inline cpp-ch/local-engine/tests/CMakeLists.txt:73]: I found benchmark_local_engine can't be built successfully with source codes in directly ch_parquet/. Need to be fixed.

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    }
+};
+
+using BuildRead = std::function<ColumnReadState(const arrow::io::ReadRange & col_range)>;
+class PageIterator;
+
+class ParquetFileReaderExt
+{
+    friend class PageIterator;
+    std::shared_ptr<::arrow::io::RandomAccessFile> source_;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:121]: Please, follow the original code style of CH codes. Class member should not end with "_".

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+    int64_t source_size_;
+    std::unique_ptr<parquet::ParquetFileReader> fileReader_;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:123]: fix style

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    std::shared_ptr<ColumnIndexFilter> column_index_filter_;
+    std::unordered_map<int32_t, std::unique_ptr<RowRanges>> row_group_row_ranges_;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:125]: Int32

File: cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp [hunk 10]
```diff
@@ -809,8 +809,11 @@ int64_t VariableLengthDataWriter::writeStruct(size_t row_idx, const DB::Tuple &
                 writer.unsafeWrite(reinterpret_cast<const char *>(&v), buffer_address + offset + start + len_null_bitmap + i * 8);
             }
             else
-                writer.unsafeWrite(
-                    reinterpret_cast<const char *>(&field_value.get<char>()), buffer_address + offset + start + len_null_bitmap + i * 8);
+            {
+                const char * src = writer.getWhichDataType().isFloat64() ? reinterpret_cast<const char *>(&field_value.get<Float64>())
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp:813]: Why?

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    std::unordered_map<int32_t, std::unique_ptr<ColumnIndexStore>> row_group_column_index_stores_;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:126]: Too much  shared_ptr and unique_ptr, use `using alias = xxx` to shorten them

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+    std::shared_ptr<parquet::internal::RecordReader> record_reader_;
+    std::unique_ptr<ParquetReadState> read_state_;
+
+    void nextRowGroup();
+    void setPageReader(std::unique_ptr<parquet::PageReader> reader, const ReadSequence & read_sequence);
+
+public:
+    VectorizedColumnReader(
+        const parquet::arrow::SchemaField & field, ParquetFileReaderExt * reader, const std::vector<int32_t> & row_groups);
+    const std::string & col_name() const { return arrowField_->name(); }
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:191]: CH style

File: cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp (hunk 10 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp:813]: Let's fix it in another pr.  It annoy me in the debug build where calling get<char> for Float64 type throw an exception.

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    bool hasMoreRead() const { return read_state_ && read_state_->hasMoreRead(); }
+    std::shared_ptr<arrow::ChunkedArray> readBatch(int64_t batch_size);
+};
+
+class VectorizedParquetRecordReader
+{
+    const DB::FormatSettings format_settings_;
+    DB::ArrowColumnToCHColumn arrowColumnToCHColumn_;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:199]: using smart prt for DB::ArrowColumnToCHColumn doesn't have an explicitly declared default constructor.

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+
+bool VectorizedParquetRecordReader::initialize(
+    const DB::Block & header,
+    const std::shared_ptr<arrow::io::RandomAccessFile> & arrow_file,
+    const std::shared_ptr<ColumnIndexFilter> & column_index_filter,
+    const std::shared_ptr<parquet::FileMetaData> & metadata)
+{
+    auto file_reader = parquet::ParquetFileReader::Open(arrow_file, parquet::default_reader_properties(), metadata);
+    const parquet::ArrowReaderProperties properties;
+    const parquet::FileMetaData & file_metadata = *(file_reader->metadata());
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:185]: why copy it.

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+    const parquet::SchemaDescriptor * parquet_schema = file_metadata.schema();
+    const auto keyValueMetadata = file_metadata.key_value_metadata();
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:187]: why copy it

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 16]
```diff
@@ -0,0 +1,523 @@
...
+    parquet::arrow::SchemaManifest manifest;
+    THROW_ARROW_NOT_OK(parquet::arrow::SchemaManifest::Make(parquet_schema, keyValueMetadata, properties, &manifest));
+    std::vector<std::shared_ptr<arrow::Field>> fields;
+    fields.reserve(manifest.schema_fields.size());
+    for (auto const & schema_field : manifest.schema_fields)
+        fields.emplace_back(schema_field.field);
+    const arrow::Schema schema(fields, keyValueMetadata);
+
+    /// column pruning
+    DB::ArrowFieldIndexUtil field_util(
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:197]: Make sure subcolumn pruning in tuple column works

Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:197]: I don't support complex type in this PR

File: cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+};
+
+using ColumnIndexInt64 = TypedColumnIndex<parquet::Int64Type>;
+using ColumnIndexInt32 = TypedColumnIndex<parquet::Int32Type>;
+
+class ColumnIndexFilter
+{
+public:
+    /// The expression is stored as Reverse Polish Notation.
+    struct RPNElement
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h:161]: What does RPN mean?

File: cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.cpp [hunk 18]
```diff
@@ -0,0 +1,982 @@
...
+    TypedComparator<DType> typed_comparator{real_value, *column_index_, *comparator_};
+    return ORDER::gt(typed_comparator);
+}
+
+template <typename DType, Derived<BoundaryOrder> ORDER>
+PageIndexs TypedColumnIndexImpl<DType, ORDER>::gtEg(const DB::Field & value) const
+{
+    auto real_value{parquet_cast<DType>(value)};
+    TypedComparator<DType> typed_comparator{real_value, *column_index_, *comparator_};
+    return ORDER::gtEq(typed_comparator);
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.cpp:595]: Is this wrong ? call `gtEq` inside `gtEg`

File: cpp-ch/local-engine/Storages/Parquet/RowRanges.h [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+        return it != ranges.end() && !it->isAfter(searchRange);
+    }
+
+    const std::vector<Range> & getRanges() const { return ranges; }
+
+    const Range & getRange(size_t index) const { return ranges[index]; }
+
+    std::string toString() const
+    {
+        std::string result = "[";
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/RowRanges.h:197]: should use `WriteBufferFromOwnString` ? Keep the taste same as `CH`

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 20]
```diff
@@ -0,0 +1,523 @@
...
+        // dictionary page header size in total_compressed_size and total_uncompressed_size
+        // (see IMPALA-694). We add padding to compensate.
+        const int64_t bytes_remaining = source_size - col_end;
+        const int64_t padding = std::min<int64_t>(kMaxDictHeaderSize, bytes_remaining);
+        col_length += padding;
+    }
+    return {col_start, col_length};
+}
+
+std::string lower_column_name_if_need(const std::string & column_name, const DB::FormatSettings & settings)
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:85]: format the name of function

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 21]
```diff
@@ -0,0 +1,523 @@
...
+}
+void emplaceReadRange(local_engine::ReadRanges & read_ranges, const int64_t offset, const int64_t length)
+{
+    if (read_ranges.empty() || !extend(read_ranges.back(), offset, length))
+        read_ranges.emplace_back(arrow::io::ReadRange{offset, length});
+}
+
+/// Compute the section of the file that should be read for the given
+/// row group and column chunk.
+::arrow::io::ReadRange ComputeColumnChunkRange(
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:52]: format the name of function

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp [hunk 22]
```diff
@@ -0,0 +1,523 @@
...
+
+    int64_t col_start = column_metadata.data_page_offset();
+    if (column_metadata.has_dictionary_page() && column_metadata.dictionary_page_offset() > 0
+        && col_start > column_metadata.dictionary_page_offset())
+        col_start = column_metadata.dictionary_page_offset();
+
+    int64_t col_length = column_metadata.total_compressed_size();
+    int64_t col_end;
+    if (col_start < 0 || col_length < 0)
+        throw parquet::ParquetException("Invalid column metadata (corrupt file?)");
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:66]: This exception may not be friendly to `Executor`. Could replace with `Exception`

File: cpp-ch/local-engine/tests/CMakeLists.txt (hunk 6 above)
Comment (baibaichen) [reply cpp-ch/local-engine/tests/CMakeLists.txt:73]: this pr fix the build

File: cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h (hunk 17 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/ColumnIndexFilter.h:161]: RPN means reverse polish notation

File: cpp-ch/local-engine/Storages/Parquet/RowRanges.h (hunk 19 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/RowRanges.h:197]: This class eventually will be replaced in Arrow,  let's keep it

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp (hunk 22 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:66]: Yes, let's fix in the follow up PRs

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp (hunk 14 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:185]: file_metadata is shared pointer

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp (hunk 15 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.cpp:187]: keyValueMetadata is shared pointer

File: cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp (hunk 10 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Parser/CHColumnToSparkRow.cpp]: add https://github.com/apache/incubator-gluten/issues/5009 for tracking this issue

File: cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h (hunk 7 above)
Comment (baibaichen) [reply cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h:121]: It's intended, It's weird that function parameter with `_`. 

We can fix them later, if it is indeed necessary to keep consistency with ch.

**Comment by lgbo-ustc** [comment]
LGTM

**Comment by baibaichen** [comment]
|              | GlutenWithCHStandard tpch-data-sf100-bucket |
| ------------ | ------------------------------------------ |
| With 4634    | [mean-total 60332, min-total 57646](https://opencicd.kyligence.com/job/gluten/job/gluten-perf/2015/)          |
| Without 4634 | [mean-total 67155, min-total 65188 ](https://opencicd.kyligence.com/job/gluten/job/gluten-perf/2014/)         |


Around 8 seconds improvment

**Comment by baibaichen** [comment]
Let's Merge

**Pull request merged**